aggregator/aggregator.go

```diff
@@ -34,116 +38,362 @@ import (
-	waitForFn                   waitForFn
-	metrics                     aggregatorMetrics
+	sync.RWMutex
+
+	opts          Options
+	nowFn         clock.NowFn
+	instanceID    string
+	shardFn       ShardFn
+	checkInterval time.Duration
+	handler       Handler
+
+	shardIDs              []uint32
+	shards                []*aggregatorShard
+	placementWatcher      services.StagedPlacementWatcher
+	placementCutoverNanos int64
+	state                 aggregatorState
+	metrics               aggregatorMetrics
+	doneCh                chan struct{}
+	wg                    sync.WaitGroup
+	sleepFn               sleepFn
 }
 
 // NewAggregator creates a new aggregator.
 func NewAggregator(opts Options) Aggregator {
-	doneCh := make(chan struct{})
+	placementWatcherOpts := opts.StagedPlacementWatcherOptions()
+	placementWatcher := placement.NewStagedPlacementWatcher(placementWatcherOpts)
 	iOpts := opts.InstrumentOptions()
-	agg := &aggregator{
-		opts:          opts,
-		nowFn:         opts.ClockOptions().NowFn(),
-		checkInterval: opts.EntryCheckInterval(),
-		doneCh:        doneCh,
-		metricMap:     newMetricMap(opts),
-		metrics:       newAggregatorMetrics(iOpts.MetricsScope(), iOpts.MetricsSamplingRate()),
+	metrics := newAggregatorMetrics(iOpts.MetricsScope(), iOpts.MetricsSamplingRate())
+
+	return &aggregator{
+		opts:                  opts,
+		nowFn:                 opts.ClockOptions().NowFn(),
+		instanceID:            opts.InstanceID(),
+		shardFn:               opts.ShardFn(),
+		checkInterval:         opts.EntryCheckInterval(),
+		handler:               opts.FlushHandler(),
+		placementWatcher:      placementWatcher,
+		placementCutoverNanos: uninitializedCutoverNanos,
+		metrics:               metrics,
+		doneCh:                make(chan struct{}),
+		sleepFn:               time.Sleep,
 	}
-	agg.addMetricWithPoliciesListFn = agg.metricMap.AddMetricWithPoliciesList
-	agg.waitForFn = time.After
+}
+
+func (agg *aggregator) Open() error {
+	agg.Lock()
+	defer agg.Unlock()
 
+	if agg.state != aggregatorNotOpen {
+		return errAggregatorIsOpenOrClosed
+	}
+	if err := agg.placementWatcher.Watch(); err != nil {
+		return err
+	}
+	if err := agg.initShardsWithLock(); err != nil {
+		return err
+	}
 	if agg.checkInterval > 0 {
 		agg.wg.Add(1)
 		go agg.tick()
 	}
-
-	agg.wg.Add(1)
-	go agg.reportMetrics()
-
-	return agg
+	agg.state = aggregatorOpen
+	return nil
 }
 
 func (agg *aggregator) AddMetricWithPoliciesList(
 	mu unaggregated.MetricUnion,
 	pl policy.PoliciesList,
 ) error {
 	callStart := agg.nowFn()
-	if atomic.LoadInt32(&agg.closed) == 1 {
+	if err := agg.checkMetricType(mu.Type); err != nil {
 		agg.metrics.addMetricWithPoliciesList.ReportError(agg.nowFn().Sub(callStart))
-		return errAggregatorClosed
+		return err
+	}
+	shard, err := agg.shardFor(mu.ID)
+	if err != nil {
+		agg.metrics.addMetricWithPoliciesList.ReportError(agg.nowFn().Sub(callStart))
+		return err
+	}
+	err = shard.AddMetricWithPoliciesList(mu, pl)
```

Comment:
Not sure I fully understand the scenario, when a placement change happens after the shard is determined:
* if the shard is still owned by the aggregator, it'll continue to write to that shard.
* if the shard is no longer owned by the aggregator or the number of shards changed, the shard will close itself asynchronously as a result, and the shard.Close function will make sure all inflight writes to that shard are completed before closing itself.